Add unique ID to compareTo calculation on Task object (closes #67)

pgbsl opened the pull request:
Fix proposal for issue #67 

I have used UUID implementation as discriminator.  Any thoughts on a better approach?

lorban commented:
You could replace the `UUID` with a simple `int` generated from a `AtomicInteger`. Other than that, this change looks good - thanks for it!

pgbsl commented:
100% agree.  I was overthinking the unique ID side of things.  Have commited with an int uniqueId, sourced from an AtomicInteger

pgbsl commented:
Ignore that last commit.  Travis build is using JDK 1.6, and the Integer.compare(int, int) method I used is JDK 1.7 feature.

I will use 1.6 code instead

lorban commented:
test this please

lorban commented:
Could you please rebase your branch and squash the four commits together since you're not using a UUID anymore to solve the bug?

Thanks!

pgbsl commented:
Squashed and rebased as requested (thanks for the travis JDK 6 fix).

Are you happy with the level of testing (i.e. the new test added to TaskSchedulerTest.java).  Happy to add in a new TaskTest class (this would probably re-use the SimpleTask defined inside TaskSchedulerTest, so I would need to promote that to a to level public class).  This could then fully test the various paths that compare to will go through

lorban commented on btm/src/test/java/bitronix/tm/timer/TaskSchedulerTest.java:
```diff
@@ -73,6 +73,19 @@ public void testTaskOrdering() throws Exception {
+
+        ts.addTask(new SimpleTask(new Date(MonotonicClock.currentTimeMillis() + 100), ts, 0, result));
+        ts.addTask(new SimpleTask(new Date(MonotonicClock.currentTimeMillis() + 200), ts, 1, result));
+        ts.addTask(new SimpleTask(new Date(MonotonicClock.currentTimeMillis() + 200), ts, 2, result));
```
There is no real guarantee that there will be two identical timestamps since you're calling `currentTimeMillis()` three times.

Storing the 'common' timestamp in a local variable then passing it to the two tasks would make this test more robust.

lorban commented:
I've added a comment in your test but other than that, it looks like this would be enough testing. Do you have other tests in mind you'd want to add?

pgbsl commented:
Yeah - sorry about that test case.  Not very robust.  Now using a shared timestamp.  I don't think that any further test cases are needed.  We could add explicit boundary tests for the compareTo method (i.e. differentTimestamps, equalTimestamps but different uniqueId, comparison with self.  But i don't think this is strictly necessary
